File: CS/PivotWithCustomColumnsAtRuntime/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Data.OleDb;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.UI.PivotGrid;
using DevExpress.XtraPivotGrid;
using DevExpress.DataAccess.Sql;
using DevExpress.DataAccess.ConnectionParameters;
using DevExpress.XtraReports.UI.CrossTab;
using DevExpress.XtraPrinting;

namespace PivotWithCustomColumnsAtRuntime {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e) {
            using (XtraReport1 report = new XtraReport1()) {

                XRCrossTab crossTab = new XRCrossTab();
                report.Detail.Controls.Add(crossTab);

                SqlDataSource ds = new SqlDataSource(new Access97ConnectionParameters(@"|DataDirectory|\nwind.mdb", String.Empty, String.Empty));
                ds.Queries.Add(SelectQueryFluentBuilder
                    .AddTable("Order Details")
                    .SelectColumns(
                        "UnitPrice",
                        "Quantity",
                        "OrderID"
                    )
                    .Join("Products", "ProductID")
                    .SelectColumns(
                        "ProductName"
                    )
                    .Filter("[Order Details].[OrderID] < 10250")
                    .Build("SalesInfo"));

                crossTab.DataSource = ds;
                crossTab.DataMember = "SalesInfo";

                crossTab.RowFields.Add(new CrossTabRowField() { FieldName = "ProductName" });
                crossTab.ColumnFields.Add(new CrossTabColumnField() { FieldName = "OrderID" });
                crossTab.DataFields.Add(new CrossTabDataField() { FieldName = "UnitPrice" });
                crossTab.DataFields.Add(new CrossTabDataField() { FieldName = "Quantity" });
                crossTab.DataFields.Add(new CrossTabDataField());
                crossTab.GenerateLayout();
                /*
                +---------------------+---------------------------------------+---------------------------------------+
                | ProductName         | [OrderID]                             | Grand total                           |
                |                     +-------------+------------+------------+-------------+------------+------------+
                |                     | Unit Price  | Quantity   | Empty cell | UnitPrice   | Quantity   | Empty cell |      
                +---------------------+-------------+------------+------------+-------------+------------+------------+
                | [ProductName]       | [UnitPrice] | [Quantity] | Empty cell | [UnitPrice] | [Quantity] | Empty cell |     
                +---------------------+-------------+------------+------------+-------------+------------+------------+                
                | Grand Total         |             |            |            |             |            |            |
                +---------------------+-------------+------------+------------+-------------+------------+------------+
                */

                //Adjust generated cells
                foreach (var c in crossTab.ColumnDefinitions) {
                    //Enable auto-width for all columns
                    c.AutoWidthMode = DevExpress.XtraReports.UI.AutoSizeMode.ShrinkAndGrow;
                }

                foreach (XRCrossTabCell c in crossTab.Cells) {
                    if (c.DataLevel == 0 && c.RowIndex != 1) {
                        //Adjust format string for the "UnitPrice" cells
                        c.TextFormatString = "{0:c}";
                    }

                    if (c.RowIndex == 0 && c.ColumnLevel == 0) {
                        //Adjust format string for the "OrderID" cells
                        c.TextFormatString = "Order {0}";
                    }

                    if (c.DataLevel == 2 && c.RowIndex != 1) {
                        //Set custom expression for empty data cells
                        c.ExpressionBindings.Add(new ExpressionBinding("Text", "[UnitPrice] * [Quantity]"));
                        c.Font = new Font(c.Font, FontStyle.Bold);
                        c.TextFormatString = "{0:c}";
                    }

                    if (c.DataLevel == 2 && c.RowIndex == 1) {
                        //Set text for empty header cells
                        c.Font = new Font(c.Font, FontStyle.Bold);
                        c.Text = "Total";
                    }                    
                }


                // Assign styles to cross tab
                crossTab.CrossTabStyles.GeneralStyle = new XRControlStyle() {
                    Name = "Default",
                    Borders = BorderSide.All,
                    Padding = new PaddingInfo() { All = 2 }
                };
                crossTab.CrossTabStyles.DataAreaStyle = crossTab.CrossTabStyles.TotalAreaStyle = new XRControlStyle() {
                    Name = "Data",
                    TextAlignment = TextAlignment.TopRight
                };
                crossTab.CrossTabStyles.HeaderAreaStyle = new XRControlStyle() {
                    Name = "HeaderAndTotals",
                    BackColor = Color.WhiteSmoke
                };

                report.ShowRibbonPreviewDialog();
            }


        }
    }
}

File: VB/PivotWithCustomColumnsAtRuntime/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Data.OleDb
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.UI.PivotGrid
Imports DevExpress.XtraPivotGrid
Imports DevExpress.DataAccess.Sql
Imports DevExpress.DataAccess.ConnectionParameters
Imports DevExpress.XtraReports.UI.CrossTab
Imports DevExpress.XtraPrinting

Namespace PivotWithCustomColumnsAtRuntime
	Partial Public Class Form1
		Inherits Form

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
			Using report As New XtraReport1()

				Dim crossTab As New XRCrossTab()
				report.Detail.Controls.Add(crossTab)

				Dim ds As New SqlDataSource(New Access97ConnectionParameters("|DataDirectory|\nwind.mdb", String.Empty, String.Empty))
				ds.Queries.Add(SelectQueryFluentBuilder.AddTable("Order Details").SelectColumns("UnitPrice", "Quantity", "OrderID").Join("Products", "ProductID").SelectColumns("ProductName").Filter("[Order Details].[OrderID] < 10250").Build("SalesInfo"))

				crossTab.DataSource = ds
				crossTab.DataMember = "SalesInfo"

				crossTab.RowFields.Add(New CrossTabRowField() With {.FieldName = "ProductName"})
				crossTab.ColumnFields.Add(New CrossTabColumnField() With {.FieldName = "OrderID"})
				crossTab.DataFields.Add(New CrossTabDataField() With {.FieldName = "UnitPrice"})
				crossTab.DataFields.Add(New CrossTabDataField() With {.FieldName = "Quantity"})
				crossTab.DataFields.Add(New CrossTabDataField())
				crossTab.GenerateLayout()
'                
'                +---------------------+---------------------------------------+---------------------------------------+
'                | ProductName         | [OrderID]                             | Grand total                           |
'                |                     +-------------+------------+------------+-------------+------------+------------+
'                |                     | Unit Price  | Quantity   | Empty cell | UnitPrice   | Quantity   | Empty cell |      
'                +---------------------+-------------+------------+------------+-------------+------------+------------+
'                | [ProductName]       | [UnitPrice] | [Quantity] | Empty cell | [UnitPrice] | [Quantity] | Empty cell |     
'                +---------------------+-------------+------------+------------+-------------+------------+------------+                
'                | Grand Total         |             |            |            |             |            |            |
'                +---------------------+-------------+------------+------------+-------------+------------+------------+
'                

				'Adjust generated cells
				For Each c In crossTab.ColumnDefinitions
					'Enable auto-width for all columns
					c.AutoWidthMode = DevExpress.XtraReports.UI.AutoSizeMode.ShrinkAndGrow
				Next c

				For Each c As XRCrossTabCell In crossTab.Cells
					If c.DataLevel = 0 AndAlso c.RowIndex <> 1 Then
						'Adjust format string for the "UnitPrice" cells
						c.TextFormatString = "{0:c}"
					End If

					If c.RowIndex = 0 AndAlso c.ColumnLevel = 0 Then
						'Adjust format string for the "OrderID" cells
						c.TextFormatString = "Order {0}"
					End If

					If c.DataLevel = 2 AndAlso c.RowIndex <> 1 Then
						'Set custom expression for empty data cells
						c.ExpressionBindings.Add(New ExpressionBinding("Text", "[UnitPrice] * [Quantity]"))
						c.Font = New Font(c.Font, FontStyle.Bold)
						c.TextFormatString = "{0:c}"
					End If

					If c.DataLevel = 2 AndAlso c.RowIndex = 1 Then
						'Set text for empty header cells
						c.Font = New Font(c.Font, FontStyle.Bold)
						c.Text = "Total"
					End If
				Next c


				' Assign styles to cross tab
				crossTab.CrossTabStyles.GeneralStyle = New XRControlStyle() With {
					.Name = "Default",
					.Borders = BorderSide.All,
					.Padding = New PaddingInfo() With {.All = 2}
				}
'INSTANT VB WARNING: An assignment within expression was extracted from the following statement:
'ORIGINAL LINE: crossTab.CrossTabStyles.DataAreaStyle = crossTab.CrossTabStyles.TotalAreaStyle = new XRControlStyle() { Name = "Data", TextAlignment = TextAlignment.TopRight };
				crossTab.CrossTabStyles.TotalAreaStyle = New XRControlStyle() With {
					.Name = "Data",
					.TextAlignment = TextAlignment.TopRight
				}
				crossTab.CrossTabStyles.DataAreaStyle = crossTab.CrossTabStyles.TotalAreaStyle
				crossTab.CrossTabStyles.HeaderAreaStyle = New XRControlStyle() With {
					.Name = "HeaderAndTotals",
					.BackColor = Color.WhiteSmoke
				}

				report.ShowRibbonPreviewDialog()
			End Using


		End Sub
	End Class
End Namespace